perf/src/recycler.rs

```diff
@@ -137,26 +328,129 @@ impl<T: Default + Reset> RecyclerX<T> {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::packet::PacketsRecycler;
+    use std::{thread::sleep, time::Duration};
 
     impl Reset for u64 {
         fn reset(&mut self) {
             *self = 10;
         }
         fn warm(&mut self, _size_hint: usize) {}
         fn set_recycler(&mut self, _recycler: Weak<RecyclerX<Self>>) {}
+        fn unset_recycler(&mut self) {}
     }
 
     #[test]
     fn test_recycler() {
-        let recycler = Recycler::default();
-        let mut y: u64 = recycler.allocate("test_recycler1");
+        let recycler = Recycler::new("");
+        let mut y: u64 = recycler.allocate().unwrap();
         assert_eq!(y, 0);
         y = 20;
         let recycler2 = recycler.clone();
         recycler2.recycler.recycle(y);
         assert_eq!(recycler.recycler.gc.lock().unwrap().len(), 1);
-        let z = recycler.allocate("test_recycler2");
+        let z = recycler.allocate().unwrap();
         assert_eq!(z, 10);
         assert_eq!(recycler.recycler.gc.lock().unwrap().len(), 0);
     }
+
+    #[test]
+    fn test_recycler_limit() {
+        let limit = 10;
+        assert!(limit <= DEFAULT_MINIMUM_OBJECT_COUNT);
+        // Use PacketRecycler so that dropping the allocated object
+        // actually recycles
+        let recycler = PacketsRecycler::new_with_limit("", limit);
+        let mut allocated_items = vec![];
+        for i in 0..limit * 2 {
+            let x = recycler.allocate();
+            if i >= limit {
+                assert!(x.is_none());
+            } else {
+                allocated_items.push(x.unwrap());
+            }
```

Comment:
maybe invert the `then` clause and `else` clause? I usually want to put something happen first in the `then` clause.